### PR TITLE
[cross] Add first version of remaining toolchain specs

### PR DESCRIPTION
**Pull request by dmcilvaney**
Add glibc stage 1, gcc stage 2, glibc stage 2, gcc stage 3.

### CODE REVIEW

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap_cross.spec [hunk 1]
```diff
@@ -149,27 +156,43 @@ which includes the C and C++ compilers.
...
 # This package contains development headers and static library for libgomp
 
 %prep
 %setup -q -n gcc-%{version}
+%patch0 -p1
 # disable no-pie for gcc binaries
 sed -i '/^NO_PIE_CFLAGS = /s/@NO_PIE_CFLAGS@//' gcc/Makefile.in
+install -vdm 755 %{_builddir}/%{name}-build
+
+install -vdm 755 %{_builddir}/%{name}-build
```
Comment (christopherco) [inline SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap_cross.spec:172]: Duplicate. Can remove

File: SPECS-CROSS/gcc-bootstrap_cross/090_all_pr55930-dependency-tracking.patch [hunk 2]
```diff
@@ -0,0 +1,18 @@
+libatomic build failure if configured with --disable-dependency-tracking
```
Comment (christopherco) [inline SPECS-CROSS/gcc-bootstrap_cross/090_all_pr55930-dependency-tracking.patch:1]: Was this patch needed to fix a build error in our gcc bootstrap phase 1, 2, and 3? Adding @vitong so he is aware.

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+%global _sbindir           /sbin
+%global _libdir            /lib
+%global _lib64dir          /lib64
+%global _libexecdir        /libexec
+%global _datadir           /share
+%global _docdir            /share/doc
+%global _includedir        /include
+%global _infodir           /share/info
+%global _mandir            /share/man
+%global _oldincludedir     /include
```
Comment (christopherco) [inline SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec:30]: How were you envisioning using `_oldincludedir` vs `_includedir`?

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+# %%files -n libgomp-devel
+# %%defattr(-,root,root)
+# %%{_cross_prefix}%%{_lib64dir}/libgomp.a
+# %%{_cross_prefix}%%{_lib64dir}/libgomp.la
+# %%{_cross_prefix}%%{_lib64dir}/libgomp.so
+# %%{_cross_prefix}%%{_lib64dir}/libgomp.spec
+
+%changelog
+* Fri Feb 12 2021 Daniel McIlvaney <damcilva@microsoft.com> - 9.1.0-11
+- Fork normal gcc package into cross compile aware boot strap ckage
```
Comment (christopherco) [inline SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec:333]:
```suggestion
- Fork normal gcc package into cross compile aware boot strap package
```
nit

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap3_cross.spec [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+Patch0:         090_all_pr55930-dependency-tracking.patch
+# Only applies to the Power9 ISA
+Patch1:         CVE-2019-15847.nopatch
+BuildRequires:  %{_cross_name}-binutils
+BuildRequires:  %{_cross_name}-kernel-headers
+BuildRequires:  %{_cross_name}-glibc-bootstrap2
+AutoReqProv:    no
+Conflicts:      %{_cross_name}-gcc
+Conflicts:      %{_cross_name}-gcc-bootstrap2
+Conflicts:      %{_cross_name}-gcc-bootstrap3
```
Comment (christopherco) [inline SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap3_cross.spec:77]: Why is there a `Conflicts` on the package itself?

File: SPECS-CROSS/gcc-bootstrap_cross/090_all_pr55930-dependency-tracking.patch (hunk 2 above)
Comment (vitong) [reply SPECS-CROSS/gcc-bootstrap_cross/090_all_pr55930-dependency-tracking.patch:1]: It looks like this patch is also used in the native gcc SPEC file. From the gnu bug link below, it seems like the bug is fixed in GCC 10 (both 10.1 and 10.2 have the fix, though I haven't tested it). We can help port the fix back to GCC 9 if that's necessary

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec (hunk 3 above)
Comment (dmcilvaney) [reply SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap2_cross.spec:30]: We need to remove this work around (actually set the /usr paths correctly when calling ./configure). See %configure in the macros file here: https://github.com/microsoft/CBL-Mariner/blob/5bc8fa1c8f0a8c8c6bd4b012a0072cbf4fd68f3e/SPECS/mariner-rpm-macros/macros#L70

File: SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap3_cross.spec (hunk 5 above)
Comment (dmcilvaney) [reply SPECS-CROSS/gcc-bootstrap_cross/gcc-bootstrap3_cross.spec:77]: Good catch, should be 1 & 2.

**Comment by vitong** [comment]
_I don't know much about SPEC files so I'm not sure if this suggestion is a good idea_

I see similar commands to set up the GCC build environment across the 3 bootstrap SPEC files. Does it make sense to have a common .sh script that the SPEC files call into in order to set up the build environment? That would somewhat reduce the amount of duplicate code.

**Comment by dmcilvaney** [comment]
Looked at refactoring to use scripts, we reference rpm macros (like %{_builddir}, %{name}, %{SOURCEx}, etc) a lot which would need to be passed to the shell script. Until this area gets more complicated I'm going to leave it directly in the .spec file.

**Review by christopherco — APPROVED** [review]

**Review by vitong — APPROVED** [review]

**Pull request merged**